### PR TITLE
[AutoDiff] Do not differentiate an 'apply' when there's no active arguments.

### PR DESCRIPTION
**Pull request by rxwei**
There are cases where an `apply` has active results while no arguments are active, specifically when the result is a buffer that is being modified by later instructions, for example:

```swift
@differentiable
func TF_529<T>(x: Vector<T>) -> Vector<T> {
  var zero = Vector<T>.zero
  zero = x
  return zero
}
```

This is now fixed by checking whether the function has any active arguments before attempting to differentiate the function.

Resolves [TF-529](https://bugs.swift.org/browse/TF-529).

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -2511,6 +2511,8 @@ static void collectMinimalIndicesForFunctionCall(
       paramIndices.push_back(currentParamIdx);
     ++currentParamIdx;
   }
+  assert((currentParamIdx > 0 || !paramIndices.empty())
```
Comment (bartchr808) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Just wanted to verify we still want this assert, or maybe change the string as there is a `"Parameter indices cannot be empty"` assert inside `visitApplyInst`

**Review by bartchr808 — APPROVED** [review]
LGTM! Just a quick question about the assert we added. :shipit: 🚢

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I just dropped it. Thanks for the catch!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**